### ISSUE TITLE
Handle "unconfigured" state of plugin gracefully

### ISSUE THREAD

**Issue by mickmister** [mattermost/mattermost-plugin-github#432]
Currently when some of the plugin settings are not provided (like the GitHub API token), the plugin bails during plugin startup. We should instead allow the plugin to start up, and show helpful errors regarding missing config values when the admin tries to set up the plugin features.

**Comment by sibasankar-demansol** [comment]
@mickmister can you let me know where exactly this issue occurred ? or step to reproduce this issue .

**Comment by mickmister** [comment]
Hi @sibasankar-demansol! Thanks for looking into this ticket

If you have not configured the plugin at all yet, installing and enabling the plugin should be all you need to do to reproduce this. The plugin will fail to start and you will have to look at the server logs to understand why it failed, which is what this ticket solves. The admin should not have to check the logs for missing configuration values.

If the plugin is already configured, you can go to the plugin's settings in the system console at http://localhost:8065/admin_console/plugins/plugin_github, and remove the github API token to purposely put the plugin into an invalid state.

In summary, having missing plugin setting config values should not cause the plugin to fail to start. We should instead gracefully go through startup, and note to the admin that they need to configure the plugin, when they try to set up a subscription or something similar.

Another thing that comes to mind is we should hide the post menu items, such as "Connect to GitHub". That should instead say "Configure GitHub plugin", and should only be visible to Mattermost sysadmins. Regular users should not see any post menu items while the plugin is not configured properly.

**Comment by mickmister** [comment]
@hanzei I'm thinking this is actually a problem with the plugin framework, not the GitHub plugin. We don't have a way for the plugin to tell the admin about missing config values, other than bailing on plugin startup and logging. Maybe we should have `OnInstallText` in the plugin manifest, or allow the error returned in `OnActivate` to be shown in the plugin management page of the system console. There's a lot of dev work to do here just because the admin doesn't know to set the config values on install.

**Comment by sibasankar-demansol** [comment]
Hi @mickmister  
i have configured thegithub plugin so removed the github api token 
do i need to restart the mattermost-server to see the logs as i didn't find the logs after removing the secret

**Comment by sibasankar-demansol** [comment]
is this the log you are referring to 
![Screenshot 2021-07-13 at 9 08 33 PM](https://user-images.githubusercontent.com/86650698/125482203-c0759f1a-a87c-4097-b6ed-9107f2aa9e40.png)

**Comment by sibasankar-demansol** [comment]
@mickmister was going through the code 
even found 2 cases 
1 .  if we are new and haven't set the web-hook secret while deploying it gets fail , which mean the plugin will not be having any control to mattermost-server , seems like it will be a major change for this , need to figure out how can we interact with server to display it over UI 
2. if i had successfully using GitHub plugin and by knowingly remover the secret while restart the server found logs the install plugin gets fail with message oAuth secret is missing , feels like here we can do some thing but still miss the major part as this error will be with mostly new users

**Comment by mickmister** [comment]
> do i need to restart the mattermost-server to see the logs as i didn't find the logs after removing the secret

@sibasankar-demansol You just need to restart the plugin, not the whole mattermost server. You can do this by toggling disable/enable in the Plugin Management page, or by saving the Enabled boolean in the github plugin settings to false, saving, then setting to true, and saving.

> this part is handled in mattermost server instead of github plugin

The `invalid config: must have a github oauth client secret` portion of this log message is coming from the GitHub plugin

https://github.com/mattermost/mattermost-plugin-github/blob/07f04cd9ee71dc327274911a9e86e30c23329c13/server/plugin/plugin.go#L123
https://github.com/mattermost/mattermost-plugin-github/blob/07f04cd9ee71dc327274911a9e86e30c23329c13/server/plugin/configuration.go#L44

If we remove this code below, then the plugin will not exit when there are config issues:

https://github.com/mattermost/mattermost-plugin-github/blob/07f04cd9ee71dc327274911a9e86e30c23329c13/server/plugin/plugin.go#L122-L124

We can instead use the `IsValid` function when the admin tries to interact with the plugin, and show an error if the config as issues.

We may want to hold off on this ticket, as it will be a considerate amount of work to get the UX correct, including the UI changes I describe in the final paragraph [here](https://github.com/mattermost/mattermost-plugin-github/issues/432#issuecomment-879133947). @hanzei Please see my other comment [here](https://github.com/mattermost/mattermost-plugin-github/issues/432#issuecomment-879135093). Do you think the UX of lazily telling the admin to configure the plugin when they try to interact with the plugin is worth the dev mana here?

**Comment by hanzei** [comment]
@mickmister I was thinking about a much simpler solution then a custom settings page:

Every slash commands check the config before running and returns an error to either contact the admin or (for admins) to configure the plugin correctly. That is a server side only implementation, which can be done in a few lines of code.

What do you think?

**Comment by mickmister** [comment]
> That is a server side only implementation, which can be done in a few lines of code.

@hanzei There is also the case of disabling/hiding the post options like "Connect to GitHub", "Create Issue", "Attach to Comment" if the plugin is not set up right. Do we want to include changes to disable these when there are plugin settings missing? I would think at least some code needs to change in the plugin's webapp for this task.

Maybe we add an endpoint that checks if the plugin is configured correctly, that is called when the "Create Issue" button etc. is clicked, and we show a generic error like "Please contact your system admin to configure the GitHub plugin." if the request returns an error. This would add some lag to opening the modal if we wait for that response before opening the modal.

**Comment by mickmister** [comment]
@hanzei Yes, I was incorrect in thinking there was also a "Connect to GitHub" post menu item, but there isn't. There is one in the Jira plugin which I assumed was also here. Yes that sounds good. We should:

- avoid returning an error in `OnActivate` if it is related to `config.IsValid`
- check for valid config on all appropriate slash commands (I think all of them), and return an error back to the client if there is a config error. This can probably be done in one spot as you implied @hanzei

**Comment by sibasankar-umpteens** [comment]
@mickmister so we should move forward with changing the code and to check isValid while every /slash command and return the error to the user

**Comment by sibasankar-umpteens** [comment]
@mickmister should we check the config valid for each and every slash command 

and this is a screenshot of slash command where aouth secret is not configured 

![Screenshot 2021-07-15 at 11 02 23 AM](https://user-images.githubusercontent.com/87431235/125734268-5258d96e-239d-4f43-820e-d44b1a1b8586.png)

**Comment by mickmister** [comment]
@sibasankar-demansol

> @mickmister so we should move forward with changing the code and to check isValid while every /slash command and return the error to the user

This is correct

> @mickmister should we check the config valid for each and every slash command

Yes every slash command

> and this is a screenshot of slash command where aouth secret is not configured
> 
> ![Screenshot 2021-07-15 at 11 02 23 AM](https://user-images.githubusercontent.com/87431235/125734268-5258d96e-239d-4f43-820e-d44b1a1b8586.png)

The error message should be more user-friendly. I think we should remove the "Encountered an error while executing command" part, and say something like:

"Before using this plugin, you will need to configure it by filling out the settings in the system console [here](INSERT_SITE_URL_HERE/admin_console/plugins/plugin_github). You can learn more about the setup process [here](https://github.com/mattermost/mattermost-plugin-github#step-3-configure-the-plugin-in-mattermost)."

The first "here" link above is `INSERT_SITE_URL_HERE/admin_console/plugins/plugin_github`, which we will have the plugin fill in the actual site URL taken from the system's settings.
